flow actions not registered in ActionRegistry — plan.choose panics

## Bug

`flow.Register(reg)` is never called outside of `internal/execution/flow/flow_test.go`. All four flow actions are unregistered in any real execution context.

### Impact

| Action | Registered? | Starlark binding? | Effect |
|--------|-------------|-------------------|--------|
| `flow.choose` | No | Yes — `plan.choose()` | **Panics** at plan time via `reg.MustGet("flow.choose")` |
| `flow.gather` | No | Yes — `plan.gather()` | Works by accident (doesn't use registry) |
| `flow.elevate` | No | No — not yet in `plan_root.go` | N/A until binding added |
| `flow.wait_until` | No | No — not yet in `plan_root.go` | N/A until binding added |

### Reproduction

```star
# Any script using plan.choose panics:
dest = t.tmp("test.txt")
plan.file.write_text(destination=dest, content="hello", mode=0o644)
exists_output = plan.file.exists(resource=dest)
plan.choose(when=exists_output, then=lambda: plan.file.remove(path=dest, prune=False, prune_boundary=""))
```

```
panic: unregistered action: flow.choose
```

### Fix

Call `flow.Register(reg)` in the test runner (`internal/e2e/testrunner/runner.go`) and in any other code path that builds an `ActionRegistry` for graph execution. The registration function exists at `internal/execution/flow/register.go` — it just isn't called.

### Affected tests (skipped pending fix)

- `test_choose_exists.star`
- `test_choose_not_exists.star`
- `test_is_dir.star`
- `test_is_file.star`

--- a/cmd/devlore-test/main.go
+++ b/cmd/devlore-test/main.go
@@ -5,62 +5,15 @@
 package main
 
 import (
-	"context"
-	"encoding/json"
-	"flag"
-	"fmt"
 	"os"
 
-	"github.com/NobleFactor/devlore-cli/internal/e2e/testrunner"
+	"github.com/NobleFactor/devlore-cli/internal/cli"
+	"github.com/NobleFactor/devlore-cli/internal/devloretest"
 )
 
 func main() {
-	script := flag.String("script", "", "path to test .star script (required)")
-	dryRun := flag.Bool("dry-run", false, "plan only, no side effects")
-	trace := flag.Bool("trace", false, "enable Starlark step trace")
-	provider := flag.String("provider", "", "restrict to a specific provider")
-	flag.Parse()
-
-	if *script == "" {
-		fmt.Fprintln(os.Stderr, "error: --script is required")
-		flag.Usage()
-		os.Exit(2)
-	}
-
-	var opts []testrunner.Option
-	if *dryRun {
-		opts = append(opts, testrunner.WithDryRun())
-	}
-	if *trace {
-		opts = append(opts, testrunner.WithTrace())
-	}
-	if *provider != "" {
-		opts = append(opts, testrunner.WithProvider(*provider))
-	}
-
-	runner := testrunner.New(*script, opts...)
-	result, err := runner.Start(context.Background())
-	if err != nil {
-		errResult := map[string]any{
-			"passed":  false,
-			"error":   err.Error(),
-			"node_count": 0,
-			"expectation_count": 0,
-			"failures": []any{},
-		}
-		data, _ := json.Marshal(errResult)
-		fmt.Println(string(data))
-		os.Exit(2)
-	}
-
-	data, err := json.Marshal(result)
-	if err != nil {
-		fmt.Fprintf(os.Stderr, "error marshaling result: %v\n", err)
-		os.Exit(2)
-	}
-	fmt.Println(string(data))
-
-	if !result.Passed {
-		os.Exit(1)
+	cmd := devloretest.NewRootCmd()
+	if err := cmd.Execute(); err != nil {
+		os.Exit(cli.ExitCode(err))
 	}
 }

--- a/internal/devloretest/cli_test.go
+++ b/internal/devloretest/cli_test.go
@@ -1,0 +1,386 @@
+// SPDX-License-Identifier: SSPL-1.0
+// Copyright (c) 2025-2026 Noble Factor. All rights reserved.
+
+package devloretest_test
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	"gopkg.in/yaml.v3"
+)
+
+// binary is the path to the compiled devlore-test binary, set by TestMain.
+var binary string
+
+// scriptPath is the absolute path to test_hello.star, set by TestMain.
+var scriptPath string
+
+func TestMain(m *testing.M) {
+	// Build the binary to a temp directory.
+	tmp, err := os.MkdirTemp("", "devlore-test-cli-*")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "creating temp dir: %v\n", err)
+		os.Exit(1)
+	}
+	defer os.RemoveAll(tmp)
+
+	binary = filepath.Join(tmp, "devlore-test")
+
+	// Find repo root (walk up from this file's directory until we find go.mod).
+	root, err := findRepoRoot()
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "finding repo root: %v\n", err)
+		os.Exit(1)
+	}
+
+	scriptPath = filepath.Join(root, "internal", "e2e", "testrunner", "data", "test_hello.star")
+
+	build := exec.Command("go", "build", "-o", binary, "./cmd/devlore-test")
+	build.Dir = root
+	build.Stderr = os.Stderr
+	if err := build.Run(); err != nil {
+		fmt.Fprintf(os.Stderr, "building devlore-test: %v\n", err)
+		os.Exit(1)
+	}
+
+	os.Exit(m.Run())
+}
+
+func findRepoRoot() (string, error) {
+	dir, err := os.Getwd()
+	if err != nil {
+		return "", err
+	}
+	for {
+		if _, err := os.Stat(filepath.Join(dir, "go.mod")); err == nil {
+			return dir, nil
+		}
+		parent := filepath.Dir(dir)
+		if parent == dir {
+			return "", fmt.Errorf("go.mod not found")
+		}
+		dir = parent
+	}
+}
+
+// run executes devlore-test with the given args and returns stdout, stderr, and exit code.
+func run(args ...string) (stdout, stderr string, exitCode int) {
+	cmd := exec.Command(binary, args...)
+	var outBuf, errBuf strings.Builder
+	cmd.Stdout = &outBuf
+	cmd.Stderr = &errBuf
+
+	err := cmd.Run()
+	exitCode = 0
+	if err != nil {
+		if ee, ok := err.(*exec.ExitError); ok {
+			exitCode = ee.ExitCode()
+		} else {
+			exitCode = -1
+		}
+	}
+	return outBuf.String(), errBuf.String(), exitCode
+}
+
+// --- Argument handling ---
+
+func TestCLI_NoArgs(t *testing.T) {
+	stdout, _, code := run()
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "Usage:")
+	assertContains(t, stdout, "devlore-test [command]")
+}
+
+func TestCLI_RunNoScript(t *testing.T) {
+	_, _, code := run("run")
+	assertExit(t, 1, code)
+}
+
+func TestCLI_RunTooManyArgs(t *testing.T) {
+	_, _, code := run("run", "a.star", "b.star")
+	assertExit(t, 1, code)
+}
+
+func TestCLI_RunMissingFile(t *testing.T) {
+	_, stderr, code := run("run", "nonexistent.star")
+	assertExit(t, 1, code)
+	assertContains(t, stderr, "no such file")
+}
+
+func TestCLI_ScriptFirst(t *testing.T) {
+	stdout, _, code := run("run", scriptPath, "--output", "receipt=/dev/null", "--output", "graph=/dev/null")
+	assertExit(t, 0, code)
+	assertValidSummary(t, stdout, true)
+}
+
+func TestCLI_ScriptMiddle(t *testing.T) {
+	stdout, _, code := run("run", "--output", "receipt=/dev/null", scriptPath, "--output", "graph=/dev/null")
+	assertExit(t, 0, code)
+	assertValidSummary(t, stdout, true)
+}
+
+func TestCLI_ScriptLast(t *testing.T) {
+	stdout, _, code := run("run", "--output", "receipt=/dev/null", "--output", "graph=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertValidSummary(t, stdout, true)
+}
+
+// --- Output routing ---
+
+func TestCLI_DefaultAllToStdout(t *testing.T) {
+	stdout, _, code := run("run", scriptPath)
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "Hello World!")
+	assertContains(t, stdout, `"passed":true`)
+	assertContains(t, stdout, "version:")
+}
+
+func TestCLI_SummaryOnly(t *testing.T) {
+	stdout, _, code := run("run", "--output", "graph=/dev/null", "--output", "receipt=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertValidSummary(t, stdout, true)
+	assertNotContains(t, stdout, "Hello World!")
+	assertNotContains(t, stdout, "version:")
+}
+
+func TestCLI_GraphOnly(t *testing.T) {
+	stdout, _, code := run("run", "--output", "summary=/dev/null", "--output", "receipt=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "Hello World!")
+	assertNotContains(t, stdout, `"passed"`)
+	assertNotContains(t, stdout, "version:")
+}
+
+func TestCLI_ReceiptOnlyYAML(t *testing.T) {
+	stdout, _, code := run("run", "--output", "graph=/dev/null", "--output", "summary=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertNotContains(t, stdout, `"passed"`)
+	assertValidYAML(t, stdout)
+	assertContains(t, stdout, "shell.exec")
+	assertContains(t, stdout, "version:")
+}
+
+func TestCLI_ReceiptOnlyJSON(t *testing.T) {
+	stdout, _, code := run("run", "--output", "graph=/dev/null", "--output", "summary=/dev/null", "--receipt-format=json", scriptPath)
+	assertExit(t, 0, code)
+	assertValidJSON(t, stdout)
+	assertContains(t, stdout, "shell.exec")
+}
+
+func TestCLI_RoutToFiles(t *testing.T) {
+	tmp := t.TempDir()
+	summaryPath := filepath.Join(tmp, "summary.json")
+	receiptPath := filepath.Join(tmp, "receipt.yaml")
+	graphPath := filepath.Join(tmp, "graph.txt")
+
+	_, _, code := run("run",
+		"--output", "summary="+summaryPath,
+		"--output", "receipt="+receiptPath,
+		"--output", "graph="+graphPath,
+		scriptPath)
+	assertExit(t, 0, code)
+
+	summary, err := os.ReadFile(summaryPath)
+	if err != nil {
+		t.Fatalf("reading summary: %v", err)
+	}
+	assertValidSummary(t, string(summary), true)
+
+	receipt, err := os.ReadFile(receiptPath)
+	if err != nil {
+		t.Fatalf("reading receipt: %v", err)
+	}
+	assertValidYAML(t, string(receipt))
+	assertContains(t, string(receipt), "shell.exec")
+
+	graph, err := os.ReadFile(graphPath)
+	if err != nil {
+		t.Fatalf("reading graph: %v", err)
+	}
+	assertContains(t, string(graph), "Hello World!")
+}
+
+func TestCLI_JSONReceiptToFile(t *testing.T) {
+	tmp := t.TempDir()
+	receiptPath := filepath.Join(tmp, "receipt.json")
+
+	_, _, code := run("run",
+		"--output", "receipt="+receiptPath,
+		"--output", "graph=/dev/null",
+		"--output", "summary=/dev/null",
+		"--receipt-format=json",
+		scriptPath)
+	assertExit(t, 0, code)
+
+	data, err := os.ReadFile(receiptPath)
+	if err != nil {
+		t.Fatalf("reading receipt: %v", err)
+	}
+	assertValidJSON(t, string(data))
+}
+
+// --- Flags ---
+
+func TestCLI_DryRun(t *testing.T) {
+	stdout, _, code := run("run", "--dry-run", "--output", "graph=/dev/null", "--output", "receipt=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertValidSummary(t, stdout, true)
+}
+
+func TestCLI_Trace(t *testing.T) {
+	stdout, _, code := run("run", "--trace", "--output", "graph=/dev/null", "--output", "receipt=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	assertContains(t, stdout, `"trace"`)
+}
+
+func TestCLI_Silent(t *testing.T) {
+	_, stderr, code := run("--silent", "run", "--output", "graph=/dev/null", "--output", "receipt=/dev/null", scriptPath)
+	assertExit(t, 0, code)
+	if stderr != "" {
+		t.Errorf("--silent should suppress stderr, got: %q", stderr)
+	}
+}
+
+// --- Error cases ---
+
+func TestCLI_InvalidOutputStream(t *testing.T) {
+	_, _, code := run("run", "--output", "bogus=/dev/null", scriptPath)
+	assertExit(t, 1, code)
+}
+
+func TestCLI_MalformedOutput(t *testing.T) {
+	_, _, code := run("run", "--output", "nodestination", scriptPath)
+	assertExit(t, 1, code)
+}
+
+func TestCLI_InvalidReceiptFormat(t *testing.T) {
+	_, _, code := run("run", "--receipt-format=xml", "--output", "receipt=/dev/null", "--output", "graph=/dev/null", scriptPath)
+	assertExit(t, 1, code)
+}
+
+func TestCLI_BadOutputPath(t *testing.T) {
+	_, _, code := run("run", "--output", "graph=/no/such/dir/out.txt", scriptPath)
+	assertExit(t, 1, code)
+}
+
+func TestCLI_UnknownFlag(t *testing.T) {
+	_, _, code := run("run", "--foobar", scriptPath)
+	assertExit(t, 1, code)
+}
+
+func TestCLI_UnknownCommand(t *testing.T) {
+	_, _, code := run("foobar")
+	assertExit(t, 1, code)
+}
+
+// --- Shared commands ---
+
+func TestCLI_Version(t *testing.T) {
+	stdout, _, code := run("version")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "Version:")
+}
+
+func TestCLI_Help(t *testing.T) {
+	stdout, _, code := run("help")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "devlore-test")
+}
+
+func TestCLI_HelpRun(t *testing.T) {
+	stdout, _, code := run("help", "run")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "<script.star>")
+}
+
+func TestCLI_ConfigPath(t *testing.T) {
+	stdout, _, code := run("config", "path")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "devlore/config.yaml")
+}
+
+func TestCLI_SelfInstallHelp(t *testing.T) {
+	stdout, _, code := run("self-install", "--help")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "--prefix")
+}
+
+func TestCLI_CompletionBash(t *testing.T) {
+	stdout, _, code := run("completion", "bash")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "bash completion")
+}
+
+func TestCLI_CompletionZsh(t *testing.T) {
+	stdout, _, code := run("completion", "zsh")
+	assertExit(t, 0, code)
+	assertContains(t, stdout, "compdef")
+}
+
+// --- Assertions ---
+
+func assertExit(t *testing.T, want, got int) {
+	t.Helper()
+	if got != want {
+		t.Errorf("exit code = %d, want %d", got, want)
+	}
+}
+
+func assertContains(t *testing.T, s, substr string) {
+	t.Helper()
+	if !strings.Contains(s, substr) {
+		t.Errorf("output missing %q (len=%d)", substr, len(s))
+	}
+}
+
+func assertNotContains(t *testing.T, s, substr string) {
+	t.Helper()
+	if strings.Contains(s, substr) {
+		t.Errorf("output should not contain %q", substr)
+	}
+}
+
+func assertValidSummary(t *testing.T, s string, wantPassed bool) {
+	t.Helper()
+	// Summary line is the first JSON line in the output.
+	for _, line := range strings.Split(strings.TrimSpace(s), "\n") {
+		line = strings.TrimSpace(line)
+		if !strings.HasPrefix(line, "{") {
+			continue
+		}
+		var result struct {
+			Passed bool `json:"passed"`
+		}
+		if err := json.Unmarshal([]byte(line), &result); err != nil {
+			t.Errorf("summary is not valid JSON: %v", err)
+			return
+		}
+		if result.Passed != wantPassed {
+			t.Errorf("summary.passed = %v, want %v", result.Passed, wantPassed)
+		}
+		return
+	}
+	t.Error("no JSON summary found in output")
+}
+
+func assertValidJSON(t *testing.T, s string) {
+	t.Helper()
+	var v any
+	if err := json.Unmarshal([]byte(strings.TrimSpace(s)), &v); err != nil {
+		t.Errorf("invalid JSON: %v", err)
+	}
+}
+
+func assertValidYAML(t *testing.T, s string) {
+	t.Helper()
+	var v any
+	if err := yaml.Unmarshal([]byte(s), &v); err != nil {
+		t.Errorf("invalid YAML: %v", err)
+	}
+}

--- a/internal/devloretest/commands.go
+++ b/internal/devloretest/commands.go
@@ -1,0 +1,190 @@
+// SPDX-License-Identifier: SSPL-1.0
+// Copyright (c) 2025-2026 Noble Factor. All rights reserved.
+
+package devloretest
+
+import (
+	"encoding/json"
+	"fmt"
+	"os"
+	"strings"
+
+	"github.com/spf13/cobra"
+	"gopkg.in/yaml.v3"
+
+	"github.com/NobleFactor/devlore-cli/internal/cli"
+	"github.com/NobleFactor/devlore-cli/internal/e2e/testrunner"
+	"github.com/NobleFactor/devlore-cli/pkg/op"
+)
+
+// outputFlags collects repeated --output key=dest flags.
+type outputFlags struct {
+	entries map[string]string
+}
+
+func (o *outputFlags) String() string {
+	var parts []string
+	for k, v := range o.entries {
+		parts = append(parts, k+"="+v)
+	}
+	return strings.Join(parts, ",")
+}
+
+func (o *outputFlags) Set(val string) error {
+	k, v, ok := strings.Cut(val, "=")
+	if !ok {
+		return fmt.Errorf("expected key=destination, got %q", val)
+	}
+	switch k {
+	case "summary", "receipt", "graph":
+		o.entries[k] = v
+		return nil
+	default:
+		return fmt.Errorf("unknown output stream %q (valid: summary, receipt, graph)", k)
+	}
+}
+
+func (o *outputFlags) Type() string {
+	return "stream=dest"
+}
+
+func newRunCmd() *cobra.Command {
+	outputs := &outputFlags{entries: map[string]string{
+		"summary": "/dev/stdout",
+		"receipt": "/dev/stdout",
+		"graph":   "/dev/stdout",
+	}}
+
+	cmd := &cobra.Command{
+		Use:   "run [flags] <script.star>",
+		Short: "Run a Starlark test script that plans and executes a graph",
+		Long: `Run a Starlark test script through the graph execution engine.
+
+The script uses plan.* bindings to build a graph and t.* assertions to
+verify expectations after execution.
+
+Three output streams can be independently routed:
+  graph    Output from the software under test
+  summary  JSON test result (passed, node_count, failures)
+  receipt  Full serialized execution graph
+
+All streams default to /dev/stdout. Route to files or /dev/null:
+  devlore-test run --output receipt=receipt.yaml test.star
+  devlore-test run --output graph=/dev/null test.star`,
+		Example: `  devlore-test run test.star
+  devlore-test run --dry-run test.star
+  devlore-test run --trace test.star
+  devlore-test run --output receipt=receipt.yaml --receipt-format=json test.star
+  devlore-test run --output graph=/dev/null --output receipt=/dev/null test.star`,
+		Args: cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			return runTest(cmd, args[0], outputs)
+		},
+	}
+
+	cmd.Flags().Bool("dry-run", false, "Plan only, no side effects")
+	cmd.Flags().Bool("trace", false, "Enable Starlark step trace")
+	cmd.Flags().String("provider", "", "Restrict to a specific provider")
+	cmd.Flags().String("receipt-format", "yaml", "Receipt format: json or yaml")
+	cmd.Flags().Var(outputs, "output", "Output routing: summary|receipt|graph=destination (repeatable)")
+
+	return cmd
+}
+
+func runTest(cmd *cobra.Command, script string, outputs *outputFlags) error {
+	dryRun, _ := cmd.Flags().GetBool("dry-run")              //nolint:errcheck // flag registered above
+	trace, _ := cmd.Flags().GetBool("trace")                 //nolint:errcheck // flag registered above
+	provider, _ := cmd.Flags().GetString("provider")         //nolint:errcheck // flag registered above
+	receiptFmt, _ := cmd.Flags().GetString("receipt-format") //nolint:errcheck // flag registered above
+
+	if receiptFmt != "json" && receiptFmt != "yaml" {
+		return fmt.Errorf("--receipt-format must be json or yaml, got %q", receiptFmt)
+	}
+
+	// Open graph output destination for streaming during execution.
+	graphOut, err := openDest(outputs.entries["graph"])
+	if err != nil {
+		return fmt.Errorf("opening graph output: %w", err)
+	}
+	defer graphOut.Close()
+
+	// Build and run.
+	var opts []testrunner.Option
+	opts = append(opts, testrunner.WithWriter(graphOut))
+	if dryRun {
+		opts = append(opts, testrunner.WithDryRun())
+	}
+	if trace {
+		opts = append(opts, testrunner.WithTrace())
+	}
+	if provider != "" {
+		opts = append(opts, testrunner.WithProvider(provider))
+	}
+
+	runner := testrunner.New(script, opts...)
+	result, err := runner.Start(cmd.Context())
+	if err != nil {
+		return err
+	}
+
+	// Write summary.
+	if err := writeSummary(outputs.entries["summary"], result); err != nil {
+		return fmt.Errorf("writing summary: %w", err)
+	}
+
+	// Write receipt.
+	if err := writeReceipt(outputs.entries["receipt"], receiptFmt, runner.Graph()); err != nil {
+		return fmt.Errorf("writing receipt: %w", err)
+	}
+
+	if !result.Passed {
+		cli.Error("test failed")
+		return cli.Failure("%d expectation(s) failed", len(result.Failures))
+	}
+
+	return nil
+}
+
+// openDest opens a destination path for writing. The caller must close the returned file.
+func openDest(path string) (*os.File, error) {
+	return os.OpenFile(path, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0o644) //nolint:gosec // G304: path from CLI flag
+}
+
+func writeSummary(dest string, result *testrunner.Result) error {
+	f, err := openDest(dest)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	data, err := json.Marshal(result)
+	if err != nil {
+		return fmt.Errorf("marshaling: %w", err)
+	}
+	_, err = fmt.Fprintln(f, string(data))
+	return err
+}
+
+func writeReceipt(dest string, format string, graph *op.Graph) error {
+	if graph == nil {
+		return nil
+	}
+
+	f, err := openDest(dest)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	switch format {
+	case "json":
+		enc := json.NewEncoder(f)
+		enc.SetIndent("", "  ")
+		return graph.Serialize(enc)
+	default:
+		enc := yaml.NewEncoder(f)
+		enc.SetIndent(2)
+		defer enc.Close()
+		return graph.Serialize(enc)
+	}
+}

--- a/internal/devloretest/root.go
+++ b/internal/devloretest/root.go
@@ -1,0 +1,117 @@
+// SPDX-License-Identifier: SSPL-1.0
+// Copyright (c) 2025-2026 Noble Factor. All rights reserved.
+
+// Package devloretest implements the devlore-test CLI commands.
+package devloretest
+
+import (
+	"fmt"
+
+	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+
+	"github.com/NobleFactor/devlore-cli/internal/cli"
+	"github.com/NobleFactor/devlore-cli/schema"
+)
+
+// Version information, set at build time via ldflags.
+var (
+	version   = "dev"
+	commit    = "none"
+	buildDate = "unknown"
+)
+
+// NewRootCmd creates the root devlore-test command with all subcommands.
+func NewRootCmd() *cobra.Command {
+	cli.SetProgramName("devlore-test")
+
+	rootCmd := &cobra.Command{
+		Use:   "devlore-test",
+		Short: "Graph test harness for Starlark plan + execute + verify",
+		Long: `devlore-test is the graph test harness for the devlore execution engine.
+
+It executes a Starlark test script that builds an execution graph, runs the
+graph through the engine, and verifies expectations against the results.
+
+Output streams:
+  graph    Output from the software under test (default: stdout)
+  summary  Test result with pass/fail and expectation counts (default: stdout)
+  receipt  Full execution graph transaction log (default: stdout)
+
+Use --output to route streams to files or /dev/null:
+  devlore-test run --output receipt=/tmp/receipt.yaml test.star
+  devlore-test run --output graph=/dev/null --output receipt=/dev/null test.star`,
+		CompletionOptions: cobra.CompletionOptions{
+			HiddenDefaultCmd: true,
+		},
+		PersistentPreRunE: func(cmd *cobra.Command, args []string) error {
+			return initConfig(cmd)
+		},
+	}
+
+	// Global flags
+	rootCmd.PersistentFlags().String("config", "", "Config file (default: ~/.config/devlore/config.yaml)")
+	rootCmd.PersistentFlags().BoolP("verbose", "v", false, "Verbose output")
+	cli.AddSilentFlag(rootCmd)
+
+	// Add subcommands
+	rootCmd.AddCommand(newRunCmd())
+
+	// Shared metadata
+	manHeader := cli.ManHeader{
+		Title:   "DEVLORE-TEST",
+		Section: "1",
+		Source:  "devlore-test " + version,
+		Manual:  "devlore-test Manual",
+	}
+	configInfo := cli.ConfigInfo{
+		Name:          "devlore-test",
+		Schema:        schema.DevloreSchema,
+		DefaultConfig: schema.TestDefaultConfig,
+	}
+
+	// Add shared commands from cli
+	rootCmd.SetHelpCommand(cli.NewHelpCmd(rootCmd, manHeader))
+	rootCmd.AddCommand(cli.NewVersionCmd(cli.VersionInfo{
+		Version:   version,
+		Commit:    commit,
+		BuildDate: buildDate,
+	}))
+	rootCmd.AddCommand(cli.NewManCmd(rootCmd, manHeader))
+	rootCmd.AddCommand(cli.NewConfigCmd(configInfo))
+	rootCmd.AddCommand(cli.NewSelfInstallCmd(rootCmd, cli.SelfInstallInfo{
+		Name:       "devlore-test",
+		ManHeader:  manHeader,
+		ConfigInfo: configInfo,
+	}))
+
+	return rootCmd
+}
+
+// initConfig initializes Viper configuration.
+func initConfig(cmd *cobra.Command) error {
+	if err := cli.InitViper(cli.ViperConfig{
+		Name:            "devlore-test",
+		EnvPrefix:       "DEVLORE_TEST",
+		UseSharedConfig: true,
+	}); err != nil {
+		return err
+	}
+
+	if cfgFile, _ := cmd.Flags().GetString("config"); cfgFile != "" { //nolint:errcheck // flag registered above
+		viper.SetConfigFile(cfgFile)
+		if err := viper.ReadInConfig(); err != nil {
+			return fmt.Errorf("failed to read config %s: %w", cfgFile, err)
+		}
+	}
+
+	if err := cli.BindFlags(cmd.Root(), "devlore-test", true); err != nil {
+		return err
+	}
+
+	if viper.GetBool("devlore-test.verbose") {
+		cli.Note("Using config: %s", viper.ConfigFileUsed())
+	}
+
+	return nil
+}

--- a/internal/e2e/testrunner/data/test_backup.star
+++ b/internal/e2e/testrunner/data/test_backup.star
@@ -1,0 +1,15 @@
+# test_backup.star — Write a file and then back it up.
+#
+# Backup moves the original to path + suffix + timestamp. We verify the
+# original is gone (backup is a move, not a copy).
+#
+# Validates: plan.file.write_text, plan.file.backup
+
+src = t.tmp("backup_src.txt")
+
+written = plan.file.write_text(destination=src, content="backup me", mode=0o644)
+plan.file.backup(path=written, backup_suffix=".bak")
+
+# Backup is a rename — the original should no longer exist.
+t.expect_no_file(src)
+t.expect_node_count(2)

--- a/internal/e2e/testrunner/data/test_choose_exists.star
+++ b/internal/e2e/testrunner/data/test_choose_exists.star
@@ -1,0 +1,23 @@
+# test_choose_exists.star — Exercise plan.choose with file.exists predicate.
+#
+# 1. Write a file
+# 2. Check existence (should be true) → choose then-branch removes it
+# 3. Check existence again (should be false) → choose is a no-op
+#
+# Validates: plan.file.write_text, plan.file.exists, plan.choose, plan.file.remove
+
+dest = t.tmp("choose_target.txt")
+
+# Step 1: Create the file.
+plan.file.write_text(destination=dest, content="delete me", mode=0o644)
+
+# Step 2: Check existence — should be true, so the then-branch fires.
+exists_output = plan.file.exists(resource=dest)
+plan.choose(
+    when=exists_output,
+    then=lambda: plan.file.remove(path=dest, prune=False, prune_boundary=""),
+)
+
+# Step 3: After removal, the file should be gone.
+t.expect_no_file(dest)
+t.expect_node_count(4)  # write_text + exists + remove + choose

--- a/internal/e2e/testrunner/data/test_choose_not_exists.star
+++ b/internal/e2e/testrunner/data/test_choose_not_exists.star
@@ -1,0 +1,21 @@
+# test_choose_not_exists.star — plan.choose with a false predicate (no then-branch).
+#
+# Check existence of a file that was never created.
+# The predicate returns false, so the then-branch should NOT execute.
+# The file we would have written in the then-branch should not exist.
+
+phantom = t.tmp("phantom.txt")
+canary  = t.tmp("canary.txt")
+
+# Predicate: file does not exist.
+exists_output = plan.file.exists(resource=phantom)
+
+# Then-branch would create canary — but it should not fire.
+plan.choose(
+    when=exists_output,
+    then=lambda: plan.file.write_text(destination=canary, content="should not exist", mode=0o644),
+)
+
+t.expect_no_file(phantom)
+t.expect_no_file(canary)
+t.expect_node_count(3)  # exists + write_text (in branch) + choose

--- a/internal/e2e/testrunner/data/test_file_lifecycle.star
+++ b/internal/e2e/testrunner/data/test_file_lifecycle.star
@@ -1,0 +1,23 @@
+# test_file_lifecycle.star — Full file lifecycle: write, copy, read back.
+#
+# 1. Write a text file
+# 2. Copy it to a second location (edge from write ensures ordering)
+# 3. Read the copy back (verifies the copy was written)
+#
+# Validates: plan.file.write_text, plan.file.copy, plan.file.read
+
+src = t.tmp("lifecycle_src.txt")
+dst = t.tmp("lifecycle_dst.txt")
+
+# Step 1: Write original.
+written = plan.file.write_text(destination=src, content="lifecycle test", mode=0o644)
+
+# Step 2: Copy to destination (edge from write ensures ordering).
+plan.file.copy(source_file=written, destination_filename=dst, destination_file_mode=0o644)
+
+# Step 3: Read the copy.
+plan.file.read(path=dst)
+
+t.expect_file(src, content="lifecycle test")
+t.expect_file(dst, content="lifecycle test")
+t.expect_node_count(3)

--- a/internal/e2e/testrunner/data/test_gather.star
+++ b/internal/e2e/testrunner/data/test_gather.star
@@ -1,0 +1,24 @@
+# test_gather.star — Use plan.gather to synchronize parallel outputs.
+#
+# Write two files independently, then gather their outputs so a
+# shell command runs only after both writes complete.
+#
+# Validates: plan.gather, plan.file.write_text, plan.shell.exec
+
+a = t.tmp("gather_a.txt")
+b = t.tmp("gather_b.txt")
+c = t.tmp("gather_c.txt")
+
+# Two independent writes.
+out_a = plan.file.write_text(destination=a, content="alpha", mode=0o644)
+out_b = plan.file.write_text(destination=b, content="bravo", mode=0o644)
+
+# Gather creates a fan-in: shell.exec depends on both writes completing.
+# The shell command concatenates both files into a third.
+g = plan.gather(out_a, out_b)
+plan.shell.exec(command="cat " + a + " " + b + " > " + c)
+
+t.expect_file(a, content="alpha")
+t.expect_file(b, content="bravo")
+t.expect_file(c, content="alphabravo")
+t.expect_node_count(3)  # write_text(a) + write_text(b) + shell.exec

--- a/internal/e2e/testrunner/data/test_hello.star
+++ b/internal/e2e/testrunner/data/test_hello.star
@@ -1,0 +1,3 @@
+# test_hello.star — Single-node graph that runs: echo "Hello World!"
+plan.shell.exec(command='echo "Hello World!"')
+t.expect_node_count(1)

--- a/internal/e2e/testrunner/data/test_is_dir.star
+++ b/internal/e2e/testrunner/data/test_is_dir.star
@@ -1,0 +1,20 @@
+# test_is_dir.star — Create a directory and use is_dir predicate in choose.
+#
+# 1. Create directory via mkdir
+# 2. Check is_dir (should be true) → then-branch writes a file inside it
+#
+# Validates: plan.file.mkdir, plan.file.is_dir, plan.choose
+
+dir  = t.tmp("is_dir_test")
+file = t.tmp("is_dir_test/proof.txt")
+
+plan.file.mkdir(resource=dir, mode=0o755)
+
+dir_check = plan.file.is_dir(resource=dir)
+plan.choose(
+    when=dir_check,
+    then=lambda: plan.file.write_text(destination=file, content="dir exists", mode=0o644),
+)
+
+t.expect_file(file, content="dir exists")
+t.expect_node_count(4)  # mkdir + is_dir + write_text + choose

--- a/internal/e2e/testrunner/data/test_is_file.star
+++ b/internal/e2e/testrunner/data/test_is_file.star
@@ -1,0 +1,20 @@
+# test_is_file.star — Write a file and use is_file predicate in choose.
+#
+# 1. Write a file
+# 2. Check is_file (should be true) → then-branch copies it
+#
+# Validates: plan.file.write_text, plan.file.is_file, plan.choose, plan.file.copy
+
+src = t.tmp("is_file_src.txt")
+dst = t.tmp("is_file_dst.txt")
+
+written = plan.file.write_text(destination=src, content="file check", mode=0o644)
+
+file_check = plan.file.is_file(resource=src)
+plan.choose(
+    when=file_check,
+    then=lambda: plan.file.copy(source_file=written, destination_filename=dst, destination_file_mode=0o644),
+)
+
+t.expect_file(dst, content="file check")
+t.expect_node_count(4)  # write_text + is_file + copy + choose

--- a/internal/e2e/testrunner/data/test_link.star
+++ b/internal/e2e/testrunner/data/test_link.star
@@ -1,0 +1,13 @@
+# test_link.star — Write a file and create a symlink to it.
+#
+# Validates: plan.file.write_text, plan.file.link
+
+target = t.tmp("link_target.txt")
+link   = t.tmp("link_pointer.txt")
+
+written = plan.file.write_text(destination=target, content="linked content", mode=0o644)
+plan.file.link(source=written, path=link)
+
+t.expect_file(target, content="linked content")
+t.expect_file(link, content="linked content")
+t.expect_node_count(2)

--- a/internal/e2e/testrunner/data/test_mkdir_and_remove_all.star
+++ b/internal/e2e/testrunner/data/test_mkdir_and_remove_all.star
@@ -1,0 +1,22 @@
+# test_mkdir_and_remove_all.star — Create a directory tree and remove it.
+#
+# 1. Create a directory with mkdir
+# 2. Write a file inside it
+# 3. Remove the entire tree with remove_all
+#
+# Validates: plan.file.mkdir, plan.file.write_text, plan.file.remove_all
+
+dir  = t.tmp("mydir")
+file = t.tmp("mydir/nested.txt")
+
+# Step 1: Create directory.
+plan.file.mkdir(resource=dir, mode=0o755)
+
+# Step 2: Write a file inside it.
+plan.file.write_text(destination=file, content="nested content", mode=0o644)
+
+# Step 3: Remove the entire directory tree.
+plan.file.remove_all(path=dir, prune=False, prune_boundary="")
+
+t.expect_no_file(file)
+t.expect_node_count(3)

--- a/internal/e2e/testrunner/data/test_move.star
+++ b/internal/e2e/testrunner/data/test_move.star
@@ -1,0 +1,13 @@
+# test_move.star — Write a file and move it to a new location.
+#
+# Validates: plan.file.write_text, plan.file.move
+
+src = t.tmp("move_src.txt")
+dst = t.tmp("move_dst.txt")
+
+written = plan.file.write_text(destination=src, content="moving data", mode=0o644)
+plan.file.move(source=written, destination=dst)
+
+t.expect_no_file(src)
+t.expect_file(dst, content="moving data")
+t.expect_node_count(2)

--- a/internal/e2e/testrunner/data/test_shell_exec.star
+++ b/internal/e2e/testrunner/data/test_shell_exec.star
@@ -1,0 +1,10 @@
+# test_shell_exec.star — Run a shell command that creates a file, verify it exists.
+#
+# Validates: plan.shell.exec with side effects visible to expectations
+
+dest = t.tmp("shell_output.txt")
+
+plan.shell.exec(command="printf 'from shell' > " + dest)
+
+t.expect_file(dest, content="from shell")
+t.expect_node_count(1)

--- a/internal/e2e/testrunner/data/test_source.star
+++ b/internal/e2e/testrunner/data/test_source.star
@@ -1,0 +1,16 @@
+# test_source.star — Use plan.source to read an existing file.
+#
+# 1. Write a file (via shell to avoid plan.file edge coupling)
+# 2. Read it back via plan.source
+#
+# Validates: plan.source (wraps file.read)
+
+dest = t.tmp("source_input.txt")
+
+# Use shell to create the file outside the graph's file provider,
+# so plan.source reads it independently.
+plan.shell.exec(command="printf 'source test' > " + dest)
+plan.source(path=dest)
+
+t.expect_file(dest, content="source test")
+t.expect_node_count(2)  # shell.exec + file.read (from source)

--- a/internal/e2e/testrunner/data/test_write_bytes.star
+++ b/internal/e2e/testrunner/data/test_write_bytes.star
@@ -1,0 +1,10 @@
+# test_write_bytes.star — Write raw bytes to a file.
+#
+# Validates: plan.file.write_bytes
+
+dest = t.tmp("bytes_output.bin")
+
+plan.file.write_bytes(destination=dest, content="raw bytes here", mode=0o644)
+
+t.expect_file(dest, content="raw bytes here")
+t.expect_node_count(1)

--- a/internal/e2e/testrunner/runner.go
+++ b/internal/e2e/testrunner/runner.go
@@ -60,6 +60,12 @@ type Runner struct {
 	trace    bool
 	provider string
 	writer   io.Writer
+	graph    *op.Graph
+}
+
+// Graph returns the execution graph after Start completes. Returns nil before Start is called.
+func (r *Runner) Graph() *op.Graph {
+	return r.graph
 }
 
 // New creates a Runner for the given script path.
@@ -98,6 +104,7 @@ func (r *Runner) Start(ctx context.Context) (*Result, error) {
 
 	// 5. Create Graph
 	graph := op.NewGraph("devlore-test")
+	r.graph = graph
 
 	// 6. Build Starlark globals
 	globals := bs.BuildGlobals(graph, "devlore-test", reg)

--- a/internal/e2e/testrunner/runner_test.go
+++ b/internal/e2e/testrunner/runner_test.go
@@ -104,6 +104,82 @@ func TestTrace(t *testing.T) {
 	}
 }
 
+func TestHello(t *testing.T) {
+	runScript(t, "test_hello.star")
+}
+
+func TestFileLifecycle(t *testing.T) {
+	runScript(t, "test_file_lifecycle.star")
+}
+
+func TestMkdirAndRemoveAll(t *testing.T) {
+	runScript(t, "test_mkdir_and_remove_all.star")
+}
+
+func TestShellExec(t *testing.T) {
+	runScript(t, "test_shell_exec.star")
+}
+
+func TestSource(t *testing.T) {
+	runScript(t, "test_source.star")
+}
+
+func TestGather(t *testing.T) {
+	runScript(t, "test_gather.star")
+}
+
+func TestMove(t *testing.T) {
+	runScript(t, "test_move.star")
+}
+
+func TestLink(t *testing.T) {
+	runScript(t, "test_link.star")
+}
+
+func TestWriteBytes(t *testing.T) {
+	runScript(t, "test_write_bytes.star")
+}
+
+func TestBackup(t *testing.T) {
+	runScript(t, "test_backup.star")
+}
+
+func TestChooseExists(t *testing.T) {
+	t.Skip("flow.choose not registered in ActionRegistry (issue #188)")
+	runScript(t, "test_choose_exists.star")
+}
+
+func TestChooseNotExists(t *testing.T) {
+	t.Skip("flow.choose not registered in ActionRegistry (issue #188)")
+	runScript(t, "test_choose_not_exists.star")
+}
+
+func TestIsDir(t *testing.T) {
+	t.Skip("flow.choose not registered in ActionRegistry (issue #188)")
+	runScript(t, "test_is_dir.star")
+}
+
+func TestIsFile(t *testing.T) {
+	t.Skip("flow.choose not registered in ActionRegistry (issue #188)")
+	runScript(t, "test_is_file.star")
+}
+
+// runScript runs a .star test script and fails on any expectation failures.
+func runScript(t *testing.T, name string) {
+	t.Helper()
+	script := filepath.Join(testdataDir(t), name)
+	runner := testrunner.New(script)
+	result, err := runner.Start(context.Background())
+	if err != nil {
+		t.Fatalf("runner error: %v", err)
+	}
+	if !result.Passed {
+		for _, f := range result.Failures {
+			t.Errorf("FAIL: %s — %s", f.Expectation, f.Message)
+		}
+	}
+}
+
 func TestDryRun(t *testing.T) {
 	script := filepath.Join(testdataDir(t), "test_write_text.star")
 	runner := testrunner.New(script, testrunner.WithDryRun())

--- a/schema/defaults/devlore-test.yaml
+++ b/schema/defaults/devlore-test.yaml
@@ -1,0 +1,7 @@
+# devlore-test Configuration
+# Location: ~/.config/devlore/config.d/devlore-test.yaml
+# Shared config: ~/.config/devlore/config.yaml
+# See: devlore-test config --schema
+
+devlore-test: {}
+  # receipt_format: yaml

--- a/schema/schema.go
+++ b/schema/schema.go
@@ -30,6 +30,12 @@ var LoreDefaultConfig []byte
 //go:embed defaults/writ.yaml
 var WritDefaultConfig []byte
 
+// TestDefaultConfig is the default devlore-test-specific configuration.
+// Installed to ~/.config/devlore/config.d/devlore-test.yaml
+//
+//go:embed defaults/devlore-test.yaml
+var TestDefaultConfig []byte
+
 // PackagesManifestSchema is the JSON schema for packages-manifest.{json,yaml} files.
 // These files declare software dependencies for writ projects.
 //

--- a/star/extensions/com.noblefactor.devlore.Test/commands/run.star
+++ b/star/extensions/com.noblefactor.devlore.Test/commands/run.star
@@ -34,7 +34,7 @@ def resolve_tool_path(ctx):
 
 def build_args(ctx, tool_path):
     """Build the devlore-test command arguments."""
-    args = [tool_path, "--script", ctx.args["script"]]
+    args = [tool_path, ctx.args["script"]]
 
     if ctx.args.get("dry-run", "false") == "true":
         args.append("--dry-run")

